server/src/main/java/org/elasticsearch/search/aggregations/pipeline/PercentileRanksBucketPipelineAggregator.java

```diff
@@ -0,0 +1,118 @@
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.pipeline;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.pipeline.BucketHelpers.GapPolicy;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class PercentileRanksBucketPipelineAggregator extends BucketMetricsPipelineAggregator {
+
+    private final double[] values;
+    private boolean keyed = true;
+    private List<Double> data;
+
+    PercentileRanksBucketPipelineAggregator(String name, double[] values, boolean keyed, String[] bucketsPaths,
+                                        GapPolicy gapPolicy, DocValueFormat formatter, Map<String, Object> metaData) {
+        super(name, bucketsPaths, gapPolicy, formatter, metaData);
+        this.values = values;
+        this.keyed = keyed;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PercentileRanksBucketPipelineAggregator(StreamInput in) throws IOException {
+        super(in);
+        values = in.readDoubleArray();
+        keyed = in.readBoolean();
+    }
+
+    @Override
+    public void innerWriteTo(StreamOutput out) throws IOException {
+        out.writeDoubleArray(values);
+        out.writeBoolean(keyed);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return PercentileRanksBucketPipelineAggregationBuilder.NAME;
+    }
+
+    @Override
+    protected void preCollection() {
+       data = new ArrayList<>(1024);
+    }
+
+    @Override
+    protected void collectBucketValue(String bucketKey, Double bucketValue) {
+        data.add(bucketValue);
+    }
+
+    public int binarySearch(List<Double> data, double target) {
+        int lo = 0;
+        int hi = data.size() - 1;
+        int mid = 0;
+        while (lo <= hi) {
+            mid = lo + (hi - lo) / 2;
+            if (data[mid] == target) return mid;
+            if (data[mid] < target) {
+                lo = mid + 1;
+            } else {
+                hi = mid - 1;
+            }
+        }
+        return lo;
+    }
+
+    @Override
+    protected InternalAggregation buildAggregation(List<PipelineAggregator> pipelineAggregators, Map<String, Object> metadata) {
+
+        // Perform the sorting and percentile rank collection now that all the data
+        // has been collected.
+        Collections.sort(data);
+        int n = data.size();
+        double[] percentileRanks = new double[values.length];
+
+        if (data.size() == 0) {
+            for (int i = 0; i < values.length; i++) {
+                percentileRanks[i] = Double.NaN;
+            }
+        } else {
+            for (int i = 0; i < values.length; i++) {
+                int index = binarySearch(data, values[i]);
```

Comment:
Ah, my fault for not being more clear.  Your implementation of binary search is great, but I think we can use the built in java `Arrays.binarySearch()`?  That way there is one less algo that we have to maintain :)

Reply:
ahh. silly mistake. Totally agree. I gave fast glance to ```Arrays.binarySearch()``` but thought it would not give me insertion point if key is not present. But looks like I was incorrect. it does provide insertion point if key not present. 

Cool! thx for catching!